core/src/G3Timesample.cxx

```diff
@@ -5,43 +5,57 @@
 
 #include <container_pybindings.h>
 #include <G3Timesample.h>
-//#include <exceptions.h>
 
-// Templates for vector operations.  g3vectype might be, for example,
-// G3VectorDouble.  One could go all the way and SFINAE this... but
-// perhaps not everything should require a black belt.
 
+// Templates for vector operations.  These work on any std::vector,
+// and thus on any G3Vector.
+
+// Returns the vector size if the type matches, and -1 otherwise.
 template <typename g3vectype>
 inline
-int g3_vect_size(const G3FrameObjectPtr vp)
+ssize_t vect_size(const G3FrameObjectPtr &vp)
 {
 	auto v = boost::dynamic_pointer_cast<const g3vectype>(vp);
 	return v == nullptr ? -1 : v->size();
 }
 
-template <typename g3vectype>
+// Concatenates two compatible vectors.
+template <typename vectype>
 inline
-void g3_concat(g3vectype &output, const g3vectype &src1, const g3vectype &src2)
+vectype vect_concat(const vectype &src1, const vectype &src2)
 {
-	auto dest = output.begin();
-	for (auto p1: src1)
-		*(dest++) = p1;
-	for (auto p2: src2)
-		*(dest++) = p2;
+	vectype output;
+	output.reserve(src1.size() + src2.size());
+	output.insert(output.end(), src1.begin(), src1.end());
+	output.insert(output.end(), src2.begin(), src2.end());
+	return output;
 }
 
+// This returns the length of the vector if it's a valid type (within
+// the context of G3Timesample).  For invalid types, it returns -1.
+inline
+ssize_t g3_vect_test_and_size(const G3FrameObjectPtr &vp)
+{
+	ssize_t vsize = -1;
+	if ((vsize = vect_size<G3VectorDouble>(vp)) < 0 &&
+	    (vsize = vect_size<G3VectorInt   >(vp)) < 0 &&
+	    (vsize = vect_size<G3VectorString>(vp)) < 0)
+		return -1;
+	return vsize;
+}
+
+// Concatenates two compatible vectors and returns the result.
+// Returns nullptr on any incompatibility.
 template <typename g3vectype>
 inline
-G3FrameObjectPtr test_and_concat(const G3FrameObjectPtr src1, const G3FrameObjectPtr src2)
+G3FrameObjectPtr test_and_concat(const G3FrameObjectPtr &src1, const G3FrameObjectPtr &src2)
```

Comment:
Looks like you're mixing pointers and references here and in the previous function.  Should use either `const G3FrameObjectPtr` or `const G3FrameObject &`, but not both.  C++ and I think Boost in particular gets super confused if you do both things at the same time.

Reply:
Passing smart pointers by const reference is well motivated for performance reasons, as it avoids reference counting operations.  This version compiles and passes tests and doesn't leak memory.  It would be surprising if we could break it so easily!

Reply:
Interesting. If it works, it works...